In [74]:
import pandas as pd
csv_files = [
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Benign_LBP.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Benign_形态学.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Benign_过滤器.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Benign_骨架特征.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Benign_高级特征.csv'
]
# 如果各CSV文件有相同的样本ID列，则保留一列即可，否则需指定合并键
merge_on_column = 'Sample_ID'  # 假设是样本ID列的名称
data_frames = []
for file_path in csv_files:
    df = pd.read_csv(file_path,header=None)
    data_frames.append(df)
Benign_df = pd.concat(data_frames, axis=1, join='outer', ignore_index=False)
Benign_df.drop(Benign_df.columns[0], axis=1, inplace=True)
Benign_df['Label'] = 0
Benign_df

1         2         3         4         5         6         7  \
0     0.058059  0.041292  0.041770  0.030273  0.023149  0.022876  0.028087   
1     0.058708  0.051248  0.027106  0.017211  0.014248  0.013115  0.014661   
2     0.041497  0.041953  0.016199  0.012814  0.013885  0.017758  0.023270   
3     0.040264  0.025661  0.030245  0.032230  0.035553  0.040689  0.048189   
4     0.040258  0.026491  0.028075  0.030627  0.034040  0.038932  0.048109   
...        ...       ...       ...       ...       ...       ...       ...   
2475  0.052696  0.033388  0.034286  0.032447  0.030568  0.031593  0.036913   
2476  0.047615  0.029783  0.040016  0.036314  0.032503  0.032559  0.038065   
2477  0.049829  0.031640  0.040739  0.037326  0.032006  0.030997  0.037161   
2478  0.041975  0.026444  0.027807  0.031494  0.033674  0.039239  0.048177   
2479  0.041242  0.026460  0.030575  0.032143  0.034320  0.038118  0.046304   

             8         9        10  ...       760       761       762  \
0     0.045512  0.053745  0.044851  ...  0.042317 -0.052999 -0.140937   
1     0.031506  0.035307  0.031559  ...  0.098308 -0.128043 -0.098877   
2     0.043000  0.064817  0.048953  ...  0.068439  0.000398  0.047242   
3     0.077196  0.083134  0.077550  ...  0.020538 -0.133970 -0.055184   
4     0.076547  0.085870  0.078065  ...  0.055476 -0.027392  0.026791   
...        ...       ...       ...  ...       ...       ...       ...   
2475  0.053742  0.061143  0.058118  ...  0.037147 -0.018968 -0.090092   
2476  0.065357  0.076245  0.070071  ...  0.083917 -0.204559  0.076698   
2477  0.059053  0.073540  0.063494  ...  0.052647  0.005318 -0.056510   
2478  0.074252  0.090720  0.074031  ...  0.031354 -0.079182 -0.115513   
2479  0.074748  0.084615  0.072202  ...  0.017500 -0.095259 -0.044797   

           763       764       765       766       767       768  Label  
0    -0.044338  0.071391 -0.020000 -0.211531 -0.201435  0.224799      0  
1    -0.031098 -0.052139 -0.169000 -0.215130 -0.104537  0.156784      0  
2     0.190759 -0.223332 -0.239000 -0.088656 -0.331340  0.067402      0  
3     0.064948  0.041458 -0.039900 -0.113489 -0.178424  0.154598      0  
4     0.144202  0.093995  0.022300 -0.194535 -0.162177  0.113191      0  
...        ...       ...       ...       ...       ...       ...    ...  
2475  0.145131  0.141736 -0.073313 -0.108215 -0.185594  0.119101      0  
2476  0.198684 -0.078079 -0.043522 -0.176732 -0.082772  0.142243      0  
2477  0.181856  0.029697  0.058450 -0.188488 -0.336352  0.100729      0  
2478  0.107716  0.059545  0.066621 -0.146356 -0.086984  0.130130      0  
2479  0.078060  0.059512 -0.010592 -0.126012 -0.091809  0.141970      0  

[2480 rows x 813 columns]

In [75]:
import pandas as pd
csv_files = [
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Malignant_LBP.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Malignant_形态学.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Malignant_过滤器.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Malignant_骨架特征.csv',
    'C:\\Users\\MYZ\\PycharmProjects\\毕设实验\\Malignant_高级特征.csv'
]
# 如果各CSV文件有相同的样本ID列，则保留一列即可，否则需指定合并键
data_frames = []
for file_path in csv_files:
    df = pd.read_csv(file_path,header=None)
    data_frames.append(df)
Malignant_df = pd.concat(data_frames, axis=1, join='outer', ignore_index=False)
Malignant_df.drop(Malignant_df.columns[0], axis=1, inplace=True)
Malignant_df['Label'] = 1
Malignant_df

1         2         3         4         5         6         7  \
0     0.064146  0.037661  0.046373  0.037845  0.029978  0.026366  0.028314   
1     0.048680  0.030807  0.038512  0.035745  0.031022  0.030432  0.035043   
2     0.065068  0.032137  0.044727  0.042317  0.036016  0.032534  0.032742   
3     0.048413  0.036463  0.038469  0.031565  0.028012  0.029435  0.037031   
4     0.051522  0.036155  0.039904  0.034481  0.030978  0.030494  0.035966   
...        ...       ...       ...       ...       ...       ...       ...   
2475  0.031755  0.019953  0.024758  0.030994  0.034981  0.041137  0.056205   
2476  0.029550  0.023065  0.019590  0.022693  0.025671  0.032891  0.050739   
2477  0.056590  0.036146  0.038090  0.033783  0.030009  0.030224  0.035295   
2478  0.055444  0.036059  0.043130  0.036509  0.030981  0.031196  0.034568   
2479  0.052534  0.035702  0.041593  0.035208  0.029963  0.029941  0.036255   

             8         9        10  ...       760       761       762  \
0     0.039540  0.039764  0.033661  ...  0.045638 -0.110468 -0.121095   
1     0.059618  0.076503  0.068363  ...  0.068195  0.007039  0.145410   
2     0.040460  0.040994  0.036941  ...  0.029324 -0.160985 -0.101699   
3     0.070019  0.082469  0.057339  ...  0.006476  0.050543 -0.116403   
4     0.059969  0.066000  0.051320  ...  0.014970  0.043391 -0.114600   
...        ...       ...       ...  ...       ...       ...       ...   
2475  0.097056  0.121208  0.090960  ...  0.077830 -0.032965 -0.064442   
2476  0.098898  0.133602  0.104019  ...  0.060514  0.043260 -0.200303   
2477  0.052807  0.056211  0.047584  ...  0.052650  0.094038  0.063539   
2478  0.052975  0.056481  0.049994  ...  0.032762 -0.071837 -0.018743   
2479  0.060280  0.065134  0.052870  ...  0.019050  0.006829 -0.034760   

           763       764       765       766       767       768  Label  
0    -0.035262  0.069131  0.016740 -0.107784 -0.137001  0.180531      1  
1     0.063756  0.029126  0.008450 -0.206205 -0.271929  0.176297      1  
2     0.089310 -0.023350  0.099837 -0.179642 -0.205744  0.107194      1  
3     0.054110 -0.103284  0.214400 -0.050038 -0.304897  0.116992      1  
4    -0.008028 -0.032253  0.144668 -0.081111 -0.231332  0.165160      1  
...        ...       ...       ...       ...       ...       ...    ...  
2475  0.039278 -0.080521  0.001011 -0.169109 -0.179915  0.046537      1  
2476  0.297538  0.045275  0.055709 -0.104880 -0.276354  0.155365      1  
2477  0.094774 -0.020944 -0.006468 -0.192807 -0.245543  0.048387      1  
2478  0.011795 -0.100116  0.101405 -0.145528 -0.305955  0.103323      1  
2479  0.042489 -0.018958  0.219072 -0.089119 -0.283946  0.057813      1  

[2480 rows x 813 columns]

In [76]:
data_df = pd.concat([Benign_df, Malignant_df], ignore_index=True)
data_df 

1         2         3         4         5         6         7  \
0     0.058059  0.041292  0.041770  0.030273  0.023149  0.022876  0.028087   
1     0.058708  0.051248  0.027106  0.017211  0.014248  0.013115  0.014661   
2     0.041497  0.041953  0.016199  0.012814  0.013885  0.017758  0.023270   
3     0.040264  0.025661  0.030245  0.032230  0.035553  0.040689  0.048189   
4     0.040258  0.026491  0.028075  0.030627  0.034040  0.038932  0.048109   
...        ...       ...       ...       ...       ...       ...       ...   
4955  0.031755  0.019953  0.024758  0.030994  0.034981  0.041137  0.056205   
4956  0.029550  0.023065  0.019590  0.022693  0.025671  0.032891  0.050739   
4957  0.056590  0.036146  0.038090  0.033783  0.030009  0.030224  0.035295   
4958  0.055444  0.036059  0.043130  0.036509  0.030981  0.031196  0.034568   
4959  0.052534  0.035702  0.041593  0.035208  0.029963  0.029941  0.036255   

             8         9        10  ...       760       761       762  \
0     0.045512  0.053745  0.044851  ...  0.042317 -0.052999 -0.140937   
1     0.031506  0.035307  0.031559  ...  0.098308 -0.128043 -0.098877   
2     0.043000  0.064817  0.048953  ...  0.068439  0.000398  0.047242   
3     0.077196  0.083134  0.077550  ...  0.020538 -0.133970 -0.055184   
4     0.076547  0.085870  0.078065  ...  0.055476 -0.027392  0.026791   
...        ...       ...       ...  ...       ...       ...       ...   
4955  0.097056  0.121208  0.090960  ...  0.077830 -0.032965 -0.064442   
4956  0.098898  0.133602  0.104019  ...  0.060514  0.043260 -0.200303   
4957  0.052807  0.056211  0.047584  ...  0.052650  0.094038  0.063539   
4958  0.052975  0.056481  0.049994  ...  0.032762 -0.071837 -0.018743   
4959  0.060280  0.065134  0.052870  ...  0.019050  0.006829 -0.034760   

           763       764       765       766       767       768  Label  
0    -0.044338  0.071391 -0.020000 -0.211531 -0.201435  0.224799      0  
1    -0.031098 -0.052139 -0.169000 -0.215130 -0.104537  0.156784      0  
2     0.190759 -0.223332 -0.239000 -0.088656 -0.331340  0.067402      0  
3     0.064948  0.041458 -0.039900 -0.113489 -0.178424  0.154598      0  
4     0.144202  0.093995  0.022300 -0.194535 -0.162177  0.113191      0  
...        ...       ...       ...       ...       ...       ...    ...  
4955  0.039278 -0.080521  0.001011 -0.169109 -0.179915  0.046537      1  
4956  0.297538  0.045275  0.055709 -0.104880 -0.276354  0.155365      1  
4957  0.094774 -0.020944 -0.006468 -0.192807 -0.245543  0.048387      1  
4958  0.011795 -0.100116  0.101405 -0.145528 -0.305955  0.103323      1  
4959  0.042489 -0.018958  0.219072 -0.089119 -0.283946  0.057813      1  

[4960 rows x 813 columns]

Virtual Adversarial Training (VAT):
输入：预提取的特征向量及其对应的标签（有标签数据）和无标签数据的特征向量。
方法：针对每个无标签数据点，计算其特征向量上的小幅度扰动（虚拟对抗扰动），使模型在此扰动下的预测分布与原特征向量的预测分布差异最大。然后在损失函数中加入一项正则化项，惩罚这种差异，促使模型在小扰动下的预测保持一致。
Mean Teacher:
输入：预提取的特征向量及其对应的标签（有标签数据）和无标签数据的特征向量。
方法：使用两套相同的模型架构，一套是学生模型（Student），用于训练；另一套是教师模型（Teacher），其权重是学生模型权重的指数移动平均（EMA）。教师模型对无标签数据的特征向量做出预测，学生模型则在数据增强（如随机翻转、裁剪等）后的特征向量上做出预测。损失函数中包含一个一致性正则化项，要求学生模型的预测尽可能接近教师模型的预测。
Temporal Ensembling:
输入：预提取的特征向量及其对应的标签（有标签数据）和无标签数据的特征向量。
方法：在训练过程中，为每个无标签数据点维护一个历史预测结果的平均（ensembling），并在每次迭代时更新这个平均。损失函数中包含一个一致性正则化项，鼓励当前模型对无标签数据特征向量的预测与该数据点的历史平均预测一致。
MixMatch:
输入：预提取的特征向量及其对应的标签（有标签数据）和无标签数据的特征向量。
方法：结合数据增强、温度平滑（temperature sharpening）和一致性正则化。对无标签数据的特征向量进行多种数据增强操作，生成多个版本。模型对这些版本做出预测，然后通过温度平滑得到伪标签。损失函数中不仅包括有标签数据的交叉熵损失，还包括无标签数据的伪标签与模型对其增强版本预测结果之间的一致性损失。
UDA (Unsupervised Data Augmentation):
输入：预提取的特征向量及其对应的标签（有标签数据）和无标签数据的特征向量。
方法：类似于MixMatch，但采用了更强的数据增强策略（如RandAugment）和更先进的标签平滑技术。模型在有标签数据和无标签数据（经过增强）上进行训练，损失函数中包含一致性正则化项，要求模型对同一无标签数据点的不同增强版本的预测保持一致。
这些方法均适用于已预提取特征向量的半监督分类任务，并通过一致性正则化策略利用无标签数据。在实际应用中，可能需要根据预提取特征向量的特性和任务需求，选择合适的数据增强策略、正则化强度以及模型架构。


我现在有一个预提取好的数据集，是dataframe格式形状为（2480*813），最后一列是标签列 请写一个适合我的数据集的Virtual Adversarial Training半监督分类代码 最好带有数据预处理过程

In [15]:
# 导入所需库
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# 加载数据集
X = data_df.iloc[:, :-1]  # 特征数据
y = data_df.iloc[:, -1]  # 目标变量（类别）

# 划分训练集和测试集
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# 创建决策树分类器
clf = DecisionTreeClassifier(random_state=42)

# 使用训练集训练模型
clf.fit(X_train, y_train)

# 预测测试集
y_pred = clf.predict(X_test)

# 计算并打印准确率
accuracy = accuracy_score(y_test, y_pred)
print(f"决策树分类器的准确率: {accuracy:.2f}")

决策树分类器的准确率: 1.00


In [6]:
# Virtual Adversarial Training (VAT)
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
# 1. 数据预处理
def load_data(data):
    data_load = data  # 替换为您的数据集路径
    X = data_load.iloc[:, :-1].values  # 特征向量
    y = data_load.iloc[:, -1].values  # 标签
    # 标准化特征向量
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)
    # 划分数据集
    X_tr, X_te, y_tr, y_te = train_test_split(X_scaled, y, test_size=0.2, random_state=42, stratify=y)
    return X_tr, X_te, y_tr, y_te

X_train, X_test, y_train, y_test = load_data(data_df) 

In [7]:
# 2. 定义自定义数据集类
class FeatureDataset(Dataset):
    def __init__(self, X, y=None, transform=None):
        self.X = torch.tensor(X, dtype=torch.float32)
        self.y = torch.tensor(y, dtype=torch.long) if y is not None else None
        self.transform = transform

    def __len__(self):
        return len(self.X)

    def __getitem__(self, idx):
        x = self.X[idx]
        y = self.y[idx] if self.y is not None else None

        if self.transform:
            x = self.transform(x)

        if y is not None:
            return x, y
        else:
            return x

In [8]:
# 3. 定义模型
class MLP(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x

In [9]:
input_dim = X_train.shape[1]
hidden_dim = 256
output_dim = len(np.unique(y_train))  
model = MLP(input_dim, hidden_dim, output_dim)

In [10]:
import torch.nn.functional as F
# 4. 定义VAT损失函数
def vat_loss(model, x, y=None, eps=1e-6, xi=1e-6):
    x_adv = x.detach() + xi * torch.randn_like(x)
    x_adv.requires_grad_()
    with torch.enable_grad():
        pred_adv = model(x_adv)
        if y is not None:
            loss = nn.CrossEntropyLoss()(pred_adv, y)
        else:
            loss = -torch.mean(torch.logsumexp(pred_adv, dim=1))
        grad = torch.autograd.grad(loss, [x_adv])[0]
        x_adv = x_adv.detach() + eps * torch.sign(grad)
        x_adv = torch.clamp(x_adv, min=x.min(), max=x.max())
    model_output = model(x)
    pred_clean = model(x)
    pred_adv = model(x_adv)
    consistency_loss = nn.KLDivLoss(reduction='batchmean')(F.softmax(pred_clean, dim=1), F.softmax(pred_adv, dim=1))

    if y is not None:
        return nn.CrossEntropyLoss()(pred_clean, y) + consistency_loss
    else:
        return consistency_loss

In [11]:
# 5. 定义训练函数
def train_epoch(model, dataloader, optimizer, device):
    model.train()
    running_loss = 0.0
    for inputs, labels in dataloader:
        inputs, labels = inputs.to(device), labels.to(device)
        optimizer.zero_grad()
        loss = vat_loss(model, inputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item() * inputs.size(0)
    return running_loss / len(dataloader.dataset)

In [12]:
# 6. 设置设备和数据加载器
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
train_dataset = FeatureDataset(X_train, y_train)
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_dataset = FeatureDataset(X_test, y_test)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)

In [13]:
# 7. 训练模型
model = model.to(device)
optimizer = optim.Adam(model.parameters(), lr=0.001)
num_epochs = 100
for epoch in range(num_epochs):
    train_loss = train_epoch(model, train_loader, optimizer, device)
    print(f'Epoch {epoch + 1}, Train Loss: {train_loss:.4f}')


Epoch 1, Train Loss: -0.4857
Epoch 2, Train Loss: -0.7111
Epoch 3, Train Loss: -0.8633
Epoch 4, Train Loss: -0.9597
Epoch 5, Train Loss: -1.0108
Epoch 6, Train Loss: -1.0356
Epoch 7, Train Loss: -1.0449
Epoch 8, Train Loss: -1.0479
Epoch 9, Train Loss: -1.0483
Epoch 10, Train Loss: -1.0480
Epoch 11, Train Loss: -1.0477
Epoch 12, Train Loss: -1.0476
Epoch 13, Train Loss: -1.0476
Epoch 14, Train Loss: -1.0477
Epoch 15, Train Loss: -1.0480
Epoch 16, Train Loss: -1.0482
Epoch 17, Train Loss: -1.0484
Epoch 18, Train Loss: -1.0487
Epoch 19, Train Loss: -1.0489
Epoch 20, Train Loss: -1.0490
Epoch 21, Train Loss: -1.0492
Epoch 22, Train Loss: -1.0493
Epoch 23, Train Loss: -1.0494
Epoch 24, Train Loss: -1.0495
Epoch 25, Train Loss: -1.0495
Epoch 26, Train Loss: -1.0496
Epoch 27, Train Loss: -1.0496
Epoch 28, Train Loss: -1.0497
Epoch 29, Train Loss: -1.0497
Epoch 30, Train Loss: -1.0498
Epoch 31, Train Loss: -1.0498
Epoch 32, Train Loss: -1.0499
Epoch 33, Train Loss: -1.0499
Epoch 34, Train Los

In [14]:
# 8. 评估模型（可选）
def evaluate(model, dataloader, device):
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for inputs, labels in dataloader:
            inputs, labels = inputs.to(device), labels.to(device)
            outputs = model(inputs)
            _, predicted = torch.max(outputs.data, 1)
            total += labels.size(0)
            correct += (predicted == labels).sum().item()

    return correct / total

test_accuracy = evaluate(model, test_loader, device)
print(f'Test Accuracy: {test_accuracy:.4f}')

Test Accuracy: 1.0000


您提供的代码实现了一个基于虚拟对抗训练（Virtual Adversarial Training, VAT）的多层感知器（Multilayer Perceptron, MLP）模型。以下是代码各部分的简要说明：
1. 数据预处理
加载数据集，提取特征向量 X 和标签 y。
使用 StandardScaler 对特征向量进行标准化。
使用 train_test_split 函数将数据划分为训练集和测试集，保持类别比例均衡。
2. 定义自定义数据集类 FeatureDataset
构建一个继承自 torch.utils.data.Dataset 的类，用于封装训练和测试数据。
存储特征向量 X 和标签 y（如果有）作为张量，并支持数据转换（如果提供）。
实现 __len__ 方法返回数据集长度，__getitem__ 方法返回指定索引处的样本及其标签（如果有）。
3. 定义 MLP 模型
创建一个继承自 torch.nn.Module 的类 MLP。
定义两个全连接层（fc1 和 fc2），分别用于特征映射和输出分类。
在 forward 方法中，先使用 ReLU 激活函数处理 fc1 的输出，再通过 fc2 得到最终的模型输出。
4. 定义 VAT 损失函数 vat_loss
计算对抗扰动 x_adv，通过在输入数据 x 上添加小幅度的随机噪声并启用梯度计算。
计算对抗样本的预测结果 pred_adv，根据是否提供标签 y，选择使用交叉熵损失或负对数似然损失。
计算对抗样本相对于损失的梯度 grad，并据此更新对抗扰动。
限制对抗扰动的范围，使其不超过原始输入数据的最小值和最大值。
计算模型在原始输入 x 和对抗扰动后的输入 x_adv 上的预测结果（pred_clean 和 pred_adv）。
使用 Kullback-Leibler 散度损失（KL divergence loss）计算一致性损失（consistency loss），衡量模型对原始输入和对抗扰动输入的预测分布差异。
最终，根据是否提供标签 y，返回交叉熵损失加上一致性损失，或者仅返回一致性损失。
5. 定义训练函数 train_epoch
在给定一个 epoch 中，对模型进行训练。
使用 model.train() 将模型置于训练模式。
初始化累计损失 running_loss 为 0。
遍历数据加载器中的每个批次，将数据和标签转移到指定设备上。
对每个批次执行以下操作：
清零优化器梯度。
计算 VAT 损失。
反向传播损失并更新模型参数。
累加当前批次的损失乘以样本数量。
返回该 epoch 的平均训练损失。
6. 设置设备和数据加载器
根据硬件条件选择使用 CPU 或 GPU。
创建训练集和测试集的 FeatureDataset 实例。
使用 DataLoader 构建训练和测试数据加载器，设置合适的批量大小和是否打乱数据。
7. 训练模型
将模型移动到指定设备。
使用 Adam 优化器初始化模型参数的学习率。
设置总的训练轮数（epochs）。
遍历所有训练轮次，对每个轮次执行以下操作：
调用 train_epoch 函数进行单轮训练。
打印当前轮次的训练损失。
综上所述，这段代码实现了基于 VAT 的 MLP 模型训练流程，包括数据预处理、自定义数据集类、模型定义、VAT 损失函数、训练函数、设备和数据加载器设置以及模型训练循环。您可以运行这段代码，根据输出的训练损失来监控模型训练过程。如果需要评估模型性能，还需在训练完成后添加相应的测试代码，计算模型在测试集上的准确率或其他评价指标。

定义评估函数 evaluate
将模型设置为评估模式（model.eval()），关闭 BatchNorm 和 Dropout 层的随机性，禁用梯度计算（with torch.no_grad():）以提高效率。
初始化正确预测数 correct 和总样本数 total 为 0。
遍历测试数据加载器中的每个批次：
将数据和标签转移到指定设备上。
通过模型计算输出 outputs。
使用 torch.max 函数找到输出张量的最大值及其对应的索引（即预测类别），取下标 0 为预测类别。
更新总样本数 total 为当前批次的样本数。
计算预测类别与真实标签之间的匹配情况，并累加正确预测数 correct。
返回正确率（correct / total），即正确预测数占总样本数的比例。
计算并打印测试准确率
调用 evaluate 函数，传入模型、测试数据加载器和设备，计算测试集上的准确率。
打印测试准确率，保留四位小数。
这段代码在模型训练完成后，能够有效地评估模型在测试集上的表现。通过调用 evaluate 函数并传入训练好的模型、测试数据加载器和设备，您可以得到模型在测试集上的准确率（Test Accuracy）。此指标可以帮助您了解模型在未见过数据上的泛化能力。如果您在训练过程中也添加了验证集，还可以用同样的方式计算验证集准确率，以便在训练过程中监控模型性能并调整超参数。